packages/core/useFetch/index.ts

```diff
@@ -276,44 +303,70 @@ export function useFetch<T>(url: MaybeRef<string>, ...args: any[]): UseFetchRetu
       }
     }
 
+    let isCanceled = false
+    const context: BeforeFetchContext = { url: unref(url), options: fetchOptions, cancel: () => { isCanceled = true } }
+
+    if (options.beforeFetch)
+      Object.assign(context, await options.beforeFetch(context))
+
+    if (isCanceled || !fetch)
+      return Promise.resolve()
+
     const defaultFetchOptions: RequestInit = {
       method: config.method,
       headers: {},
     }
 
-    if (config.payload) {
+    let payload = context.options?.body || config.payload
+
+    if (payload) {
       const headers = defaultFetchOptions.headers as Record<string, string>
       if (config.payloadType === 'json') {
-        defaultFetchOptions.body = JSON.stringify(config.payload)
+        payload = JSON.stringify(payload)
         headers['Content-Type'] = 'application/json'
       }
       else {
-        defaultFetchOptions.body = config.payload as any
-        headers['Content-Type'] = config.payloadType === 'formData'
-          ? 'multipart/form-data'
-          : 'text/plain'
+        if (config.payloadType === 'formData' || payload instanceof FormData) {
+          // we just add it here to remove later once merged with context.options.headers
+          headers['Content-Type'] = 'multipart/form-data'
+        }
+        else if (config.payloadType === 'text') {
+          headers['Content-Type'] = 'text/plain'
+        }
+        else {
+          headers['Content-Type'] = 'application/x-www-form-urlencoded'
+          // if the payload is not URLSearchParams nor a string, we need to encode it
+          // if it is a string it should be already encoded: avoiding double encoding
+          if (!(payload instanceof URLSearchParams) && typeof payload !== 'string')
```

Comment:
Wondering if we shouldn't leave this up to the user to do. I'm not 100% if we want to be modifying the payload. We could always provide a separate function that the user could use to encode it instead. Also the fetch function is getting quite large and since this is something that probably won't be used a lot which makes me think even more that we can just provide an encode function instead

Reply:
I think none of this code is needed, not even as a separate function. `fetch` works fine even with `application/x-www-form-urlencoded` and `multipart/form-data`.

This will set content type to `application/x-www-form-urlencoded;charset=UTF-8`
```js
const res = await fetch('server.com/blabla', {
    method: 'post',
    body: new URLSearchParams({ email: 'email@email.com', password: 'asdasd' })
});
```

This to `multipart/form-data; boundary...`
```js
const fd = new FormData();
fd.append('email', 'email@e.com');
fd.append('password', 'passw');
const res = await fetch('server.com/blabla', {
    method: 'post',
    body: fd
});
```
Strings will map to `text/plain;`.

The problem with `useFetch` is that it's setting the type to `text` for body other then typeof object. Why not set the content type only if no `payloadType` was provided and the payload is of type object. For the rest you just pass them as they are.
Also this part can be updated before making the call
```diff
else {
        defaultFetchOptions.body = config.payload;
-        headers['Content-Type'] = config.payloadType === 'formData'
-               ? 'multipart/form-data'
-               : 'text/plain';
 }
```
No code is better code.

Reply:
We are not modifying the payload, we are encoding the payload, this is equivalent to `JSON.stringnify` when using json payload. Just check the devtools of the payload sent in demo page, see original data on the network tab, just include &%<> in any form field.
If the data must be encoded by the user, what can we do if just providing an object?

Reply:
I think the API provided by `useFetch` should be simpler, if the payload is an object the content type will be set to `application/json` and the payload will be stringified, else payload will go to `fetch` as is also the content type if provided. This way these should work:
```js
// This will set the type to json, `fetch` does not accept an object as body so we suppose
// it's intended to be used as json
const {data} = useFetch('/api').post({data: 'some data'})
```
```js
// Content type is set to `application/x-www-form-urlencoded` and properly encoded
// `useFetch` should just pass the payload and it works
// The same will be true also with `multipart/form-data` and `new FormData()`
const {data} = useFetch('/api').post(new URLSearchParams({data: 'some data &%+ '}))
```

I would strongly recommend the use of `URLSearchParams` and `FormData` instead of custom encoding functions.

Reply:
Then, why are you encoding when using `json` payload? The implementation should be consistent, `useFetch` should deal with all encodings or none of them.

In the other hand, using a `multipart/form-data` to submit a form? This will have server side effects, mappings should be added/modified to just handle a `multipart/form-data`.

I have changed the `application/x-www-form-urlencoded` to build a `URLSearchParams` from the body object and so we don't need to encoded it explicitly. The method added `toURLSearchParams` can be exported, right now is internal.

About the `payloadType`, in this PR I have already change how we must handle it. In fact this was the cause I make this PR: the payload handling should check both, the `payloadType` and the `payload` itself, since the user can provide incoherent configuration and payload.

Reply:
> Then, why are you encoding when using `json` payload? The implementation should be consistent, useFetch should deal with all encodings or none of them

Because only `object` can't be in body, `string`, `URLSearchParams` and `FormData` can be and the content type is handled by default by `fetch`.

And yes I'd prefer `useFetch` to be a thin wrapper to `fetch` which adds only the reactivity part without touching the payload, modifying the header or __throwing an error when status is not ok__ (I really don't like this behaviour :shrug:). If someone wants these behaviours they can use:
```js
const useCustomFetch = createFetch({
    beforeFetch() {/*do what you think is ok, encode, stringify ...*/},
    afterFetch() {/*you' can throw here if you want*/}
})

Reply:
What do you mean with modifying the payload? `useFecth` is not modifiying the payload, it is encoding to what `fetch` expects.

Take a look at #542, the method building the body uses the `content-type` header provided by the user. Since `useFetch` has `payloadType` configuration option, it is a pain, since we must deal with `fetch` on `multipart/form-data` and `application/x-www-form-urlencoded` and the type of the payload (like original does with `json` payload). 

Just comment lines https://github.com/vueuse/vueuse/blob/37221e486ba718bd7b6c10949b8e3a2892ebc5f1/packages/core/useFetch/index.ts#L341 and https://github.com/vueuse/vueuse/blob/37221e486ba718bd7b6c10949b8e3a2892ebc5f1/packages/core/useFetch/index.ts#L342, start the app, go to demo page, and send the multipart, the files entry from the server response will be missing.

We need to remove the header for `multipart/form-data`, since we are breaking `fetch`: it will be not possible to parse the multipart on server (once set `content-type`, the parts that `fetch` add to the request are not correlated to this header: it will change between executions, and if configured, that is, not removed, then it will be fixed).

Reply:
Commeting out the lines result in:

![imagen](https://user-images.githubusercontent.com/6311119/120631237-f1108780-c467-11eb-8640-3e075eef412a.png)

and the response from server result in:

![imagen](https://user-images.githubusercontent.com/6311119/120631336-0eddec80-c468-11eb-9f33-74cac9030701.png)

while removing the header will be:

![imagen](https://user-images.githubusercontent.com/6311119/120631472-36cd5000-c468-11eb-9a53-4b43786fc3c5.png)

with the response from server:

![imagen](https://user-images.githubusercontent.com/6311119/120631517-4482d580-c468-11eb-927a-515b0f9b2186.png)

Reply:
The problem here is the boundary added to the `content-type` header by `fetch`, that cannot be configure on the parts...

Reply:
We also need to avoid double encoding on `json` payload:
```javascript
const x = JSON.stringify({a: 1, b: 2})
const y = JSON.stringify(x)
```
```console
"{\"a\":1,\"b\":2}"
"\"{\\\"a\\\":1,\\\"b\\\":2}\""
```

Reply:
> What do you mean with modifying the payload? useFecth is not modifiying the payload, it is encoding to what fetch expects.

But why handle some cases and not let the user decide how the data is sent? `fetch` will handle most of the cases a part from the `object`. For consistency we ignore this fact and let the user stringify the body and set the content type. Or since json is a common type, `useFetch` does the encoding only in this particular case, the rest is up to the user.

> Since `useFetch` has `payloadType` configuration option, it is a pain

Yes, it has a problem and it is a pain. That's why I'm saying `useFetch` shouldn't add `content-type` or deal with that.

> The problem here is the boundary added to the `content-type` header by `fetch`

Exactly, don't deal with content-type, do not add and remove it. It will work like `post(new FormData())` for multipart, `post(new URLSearchParams())` for url encoded.  It's easy, it's up to the user

Reply:
I think the user expects that passing an object, `useFetch` can deal with it via `payloadType`. How many lines you must write to use `URLSearchParams` or `FormData` while if using a simple object (see demo page), the user can forget about the encoding logic?.

In fact, the same approach can be done for `multipart/form-data`, that is, the user provides a raw object, and `useFetch` can deal with it: create a  `FormData` adding each key to the  `FormData`, and if a `File` entry then configure also the name.

As a said before, deal with all encodings or none of them: why `json` type and not `formEncoded` or `formData`? Because the simplicity of the conversion?

Reply:
Of course, if the user provides the object `fetch` expects, `useFetch` will do nothing... 

Another story would be if we had to square the circle to make the encodings, but if it is 2 lines of code!!!.

Reply:
I think the behavior I am trying to add to `useFetch` is good from a user usage perspective.

The number of lines that we save and the abstraction in the encoding that we will provide to the users will be huge.

Give `useFetch` an object and tell him which encoding you want to apply, which will take care of what it has to do to send it to the server. 

This is how I think `useFetch` should work.

Reply:
I'll push changes to build the multipart from a raw object.

This PR will have the code with the functionality I think `useFetch` should have.

There is one thing left to do about error handling, @isgj can you see my comments on #553?

We can have later a second round...

Reply:
I'll provide some docs here, once reviewed last changes I'll add to the docs...